src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs

```diff
@@ -1803,188 +1803,270 @@ public static bool Decompose(Matrix4x4 matrix, out Vector3 scale, out Quaternion
-                            {
-                                cc = 0;
-                            }
-                            else
-                            {
-                                if (fAbsX < fAbsZ)
-                                {
-                                    cc = 0;
-                                }
-                                else
-                                {
-                                    cc = 2;
-                                }
-                            }
+                            canonicalBasis_CC = canonicalBasis_Row1;
                         }
                         else
                         {
-                            if (fAbsX < fAbsZ)
-                            {
-                                cc = 1;
-                            }
-                            else
-                            {
-                                if (fAbsY < fAbsZ)
-                                {
-                                    cc = 1;
-                                }
-                                else
-                                {
-                                    cc = 2;
-                                }
-                            }
+                            canonicalBasis_CC = canonicalBasis_Row2;
                         }
-                        #endregion
-
-                        *pVectorBasis[b] = Vector3.Cross(*pVectorBasis[a], *(pCanonicalBasis + cc));
                     }
+                }
+                #endregion
 
-                    *pVectorBasis[b] = Vector3.Normalize(*pVectorBasis[b]);
-
-                    if (pfScales[c] < DecomposeEpsilon)
-                    {
-                        *pVectorBasis[c] = Vector3.Cross(*pVectorBasis[a], *pVectorBasis[b]);
-                    }
+                pVectorBasis_B = Vector3.Cross(pVectorBasis_A, canonicalBasis_CC);
+            }
 
-                    *pVectorBasis[c] = Vector3.Normalize(*pVectorBasis[c]);
+            pVectorBasis_B = Vector3.Normalize(pVectorBasis_B);
 
-                    det = matTemp.GetDeterminant();
+            if (pfScales_C < DecomposeEpsilon)
+            {
+                pVectorBasis_C = Vector3.Cross(pVectorBasis_A, pVectorBasis_B);
+            }
 
-                    // use Kramer's rule to check for handedness of coordinate system
-                    if (det < 0.0f)
-                    {
-                        // switch coordinate system by negating the scale and inverting the basis vector on the x-axis
-                        pfScales[a] = -pfScales[a];
-                        *pVectorBasis[a] = -(*pVectorBasis[a]);
+            pVectorBasis_C = Vector3.Normalize(pVectorBasis_C);
+            if (a == 0)
+            {
+                Unsafe.As<float, Vector3>(ref matTemp.M11) = pVectorBasis_A;
+                if (b == 1)
+                {
+                    Unsafe.As<float, Vector3>(ref matTemp.M21) = pVectorBasis_B;
```

Comment:
`matTemp` is a local matrix, so its on the stack and already pinned. Using pointers would be better and smaller IL.